### PR TITLE
[serve] Add api ref and telemetry for new multi-app api

### LINKED ISSUE #38015: [serve] Add telemetry for new multi-app API
Add telemetry for new Serve API `serve.status()` and `serve.get_app_handle`/`serve.get_deployment_handle`.

### PR DESCRIPTION
**Pull request by zcin**
## Why are these changes needed?

Add `serve.status()`, `serve.get_app_handle()` and `serve.get_deployment_handle()` to Serve api refs.
Add telemetry for `serve.status`, `serve.get_app_handle`, `serve.get_deployment_handle`

## Related issue number

Closes https://github.com/ray-project/ray/issues/38015

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/protobuf/usage.proto
Comment (GeneDer) [inline src/ray/protobuf/usage.proto:70]: I know there could be some merge conflicts with other PRs if you start with 18. But I feel we shouldn't skip numbers until then. Whichever PR merged later should resolve the conflicts. Also, would be nice if we can add comments like `("1" if used)`

File: doc/source/serve/api/index.md
Comment (GeneDer) [inline doc/source/serve/api/index.md:70]: Since there are different sets of code owners, I feel it might be better to just split this into it's own PR, but no biggie

File: python/ray/serve/tests/test_telemetry.py
Comment (GeneDer) [inline python/ray/serve/tests/test_telemetry.py:511]: Are we intended to add a parameter to test when it's not called? If not, maybe we can just put a check before the if statement to ensure they have yet exist in the report :) Same goes to other ones.

Comment (zcin) [reply python/ray/serve/tests/test_telemetry.py:511]: Ah that was meant to be removed

File: src/ray/protobuf/usage.proto
Comment (zcin) [reply src/ray/protobuf/usage.proto]: Good point, I've changed these to start at 18!

File: doc/source/serve/api/index.md
Comment (zcin) [reply doc/source/serve/api/index.md:70]: Good point, will keep in mind for future PRs!

File: doc/source/serve/api/index.md
Comment (edoakes) [inline doc/source/serve/api/index.md:56]: add `get_deployment_handle` too

make sure the docstring clarifies that it is a developer API & any limitations

**Review by edoakes — APPROVED** [review]

File: python/ray/serve/tests/test_telemetry.py
Comment (GeneDer) [inline python/ray/serve/tests/test_telemetry.py:511]: Non-blocker: I think it would be nice to check those telemetries doesn't exist before the api calls just to double check (pretty sure it will pass as well). Same goes to other ones.

**Review by GeneDer — APPROVED** [review]
Minor comment on the test not blocker :)

File: python/ray/serve/tests/test_telemetry.py
Comment (zcin) [reply python/ray/serve/tests/test_telemetry.py:505]: Makes sense, I've added it as a test case to all 3 tests!

File: doc/source/serve/api/index.md
Comment (zcin) [reply doc/source/serve/api/index.md:56]: Added and clarified in the docstring!

File: python/ray/serve/tests/test_telemetry.py
Comment (edoakes) [inline python/ray/serve/tests/test_telemetry.py:114]: this is not acceptable, it will lead to subtle bugs in the future. please try to re-write the test in a way that it does not produce side effects

**Review by edoakes — CHANGES_REQUESTED** [review]

**Review by edoakes — APPROVED** [review]

**Comment by zcin** [comment]
@edoakes this change is ready for merge.

**Comment by edoakes** [comment]
@zcin needed to merge master due to failing required build

File: python/ray/serve/tests/test_telemetry.py
Comment (edoakes) [inline python/ray/serve/tests/test_telemetry.py:54]: you can use the `ServeUsageKey.get_value_from_report` method for this (check `is None`)

File: python/ray/serve/api.py
Comment (angelinalg) [inline python/ray/serve/api.py:841]:
```suggestion
    This is a developer API and is for advanced Ray
```

**Review by angelinalg — APPROVED** [review]

**Comment by angelinalg** [comment]
Thanks for updating the API refs!

**Review by thomasdesr — APPROVED** [review]

**Comment by zcin** [comment]
Separated `test_telemetry` into `test_telemetry` and `test_telemetry_2` because it was timing out on Windows.

**Comment by zhe-thoughts** [comment]
@edoakes Could you signoff for `tests-ok`? Thanks

**Review by zhe-thoughts — APPROVED** [review]
Agreed this qualifies for merging. But waiting for `tests-ok` green light

**Comment by edoakes** [comment]
@zhe-thoughts tagged

**Pull request merged**
